File: knowhub/src/main/java/com/knowhub/pojo/resource/entity/Resource.java
package com.knowhub.pojo.resource.entity;

import com.rookie.common.pojo.BaseEntity;

import java.util.Date;

/**
 * 资源主表实体，对应 resource 表。
 * 资源分两类：FILE 文件（走文件存储模块上传，关联 file_object_id）/ LINK 链接（存 link_url）。
 * 主表只存状态机字段(status/review_status/publish_time)，审核员/审核时间/审核意见全在
 * resource_review_log 流水表（不冗余主表快照，比博客主表更干净）。
 * 互动计数(点赞/收藏/评分)不冗余主表，走事实表聚合；仅 download_count 冗余主表(仅FILE下载+1)。
 * authorId 为作者用户ID(userId)，与审计列 createBy(username) 互补：
 *   createBy 存账号串便于直接展示，authorId 用 userId 稳定锁定（username 可改，userId 不变），
 *   前台展示作者昵称走 join sys_user on user_id=author_id，username 变动不影响。
 * resourceCategoryId 写全名（不写 categoryId，防歧义），-1=其他（前端硬编码约定）。
 * 审计列(createBy/updateBy/createTime/updateTime) 由 BaseEntity 承载；软删 deleted 独立字段。
 */
public class Resource extends BaseEntity {

    private Long resourceId;

    private Long authorId;

    /** 资源类型：FILE 文件 / LINK 链接（见 ResourceType 枚举） */
    private String resourceType;

    /** 分类 id，-1=其他（前端硬编码约定，不查分类表） */
    private Long resourceCategoryId;

    private String title;

    private String summary;

    /** 详细说明（支持 Markdown） */
    private String description;

    /** FILE 类型：关联 file_object.object_id（RESOURCE_FILE 业务） */
    private Long fileObjectId;

    /** LINK 类型：外部链接 URL */
    private String linkUrl;

    /** LINK 类型：图标 URL（首版运行时拼 favicon，可空） */
    private String linkIcon;

    /** 状态：DRAFT/PUBLISHED/REVOKED/PENDING_REVIEW/REJECTED（见 ResourceStatus 枚举） */
    private String status;

    /** 审核状态：NONE/PENDING/APPROVED/REJECTED（见 ReviewStatus 枚举，复用） */
    private String reviewStatus;

    private Date publishTime;

    /** 下载次数（仅 FILE 下载 +1，LINK 点击不计） */
    private Long downloadCount;

    /** 浏览量（独立访客数，user_view_history 首次 INSERT +1，冗余列读快；与 download_count 正交） */
    private Long viewCount;

    private Integer deleted;

    // ---- 非表字段（列表/详情查询 join 带出的展示字段，resultMap 映射，不入库） ----
    /** 作者昵称（join sys_user on user_id=author_id 带出，非表字段） */
    private String authorNickname;

    /** 分类名（join resource_category 带出；-1=其他时为 null，非表字段） */
    private String categoryName;

    /** FILE 类型：原始文件名（join file_object 带出，非表字段） */
    private String originalName;

    /** FILE 类型：文件大小字节（join file_object 带出，非表字段） */
    private Long contentLength;

    /** FILE 类型：MIME 类型（join file_object 带出，非表字段） */
    private String contentType;

    public Resource() {
    }

    public Resource(Date createTime, Date updateTime, String createBy, String updateBy,
                    Long resourceId, Long authorId, String resourceType, Long resourceCategoryId,
                    String title, String summary, String description,
                    Long fileObjectId, String linkUrl, String linkIcon,
                    String status, String reviewStatus, Date publishTime, Long downloadCount, Integer deleted) {
        super(createTime, updateTime, createBy, updateBy);
        this.resourceId = resourceId;
        this.authorId = authorId;
        this.resourceType = resourceType;
        this.resourceCategoryId = resourceCategoryId;
        this.title = title;
        this.summary = summary;
        this.description = description;
        this.fileObjectId = fileObjectId;
        this.linkUrl = linkUrl;
        this.linkIcon = linkIcon;
        this.status = status;
        this.reviewStatus = reviewStatus;
        this.publishTime = publishTime;
        this.downloadCount = downloadCount;
        this.deleted = deleted;
    }

    public Long getResourceId() {
        return resourceId;
    }

    public void setResourceId(Long resourceId) {
        this.resourceId = resourceId;
    }

    public Long getAuthorId() {
        return authorId;
    }

    public void setAuthorId(Long authorId) {
        this.authorId = authorId;
    }

    public String getResourceType() {
        return resourceType;
    }

    public void setResourceType(String resourceType) {
        this.resourceType = resourceType;
    }

    public Long getResourceCategoryId() {
        return resourceCategoryId;
    }

    public void setResourceCategoryId(Long resourceCategoryId) {
        this.resourceCategoryId = resourceCategoryId;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getSummary() {
        return summary;
    }

    public void setSummary(String summary) {
        this.summary = summary;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Long getFileObjectId() {
        return fileObjectId;
    }

    public void setFileObjectId(Long fileObjectId) {
        this.fileObjectId = fileObjectId;
    }

    public String getLinkUrl() {
        return linkUrl;
    }

    public void setLinkUrl(String linkUrl) {
        this.linkUrl = linkUrl;
    }

    public String getLinkIcon() {
        return linkIcon;
    }

    public void setLinkIcon(String linkIcon) {
        this.linkIcon = linkIcon;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public String getReviewStatus() {
        return reviewStatus;
    }

    public void setReviewStatus(String reviewStatus) {
        this.reviewStatus = reviewStatus;
    }

    public Date getPublishTime() {
        return publishTime;
    }

    public void setPublishTime(Date publishTime) {
        this.publishTime = publishTime;
    }

    public Long getDownloadCount() {
        return downloadCount;
    }

    public void setDownloadCount(Long downloadCount) {
        this.downloadCount = downloadCount;
    }

    public Long getViewCount() {
        return viewCount;
    }

    public void setViewCount(Long viewCount) {
        this.viewCount = viewCount;
    }

    public Integer getDeleted() {
        return deleted;
    }

    public void setDeleted(Integer deleted) {
        this.deleted = deleted;
    }

    public String getAuthorNickname() {
        return authorNickname;
    }

    public void setAuthorNickname(String authorNickname) {
        this.authorNickname = authorNickname;
    }

    public String getCategoryName() {
        return categoryName;
    }

    public void setCategoryName(String categoryName) {
        this.categoryName = categoryName;
    }

    public String getOriginalName() {
        return originalName;
    }

    public void setOriginalName(String originalName) {
        this.originalName = originalName;
    }

    public Long getContentLength() {
        return contentLength;
    }

    public void setContentLength(Long contentLength) {
        this.contentLength = contentLength;
    }

    public String getContentType() {
        return contentType;
    }

    public void setContentType(String contentType) {
        this.contentType = contentType;
    }

    @Override
    public String toString() {
        return "Resource{" +
                "resourceId=" + resourceId +
                ", authorId=" + authorId +
                ", resourceType='" + resourceType + '\'' +
                ", resourceCategoryId=" + resourceCategoryId +
                ", title='" + title + '\'' +
                ", status='" + status + '\'' +
                ", reviewStatus='" + reviewStatus + '\'' +
                ", publishTime=" + publishTime +
                ", downloadCount=" + downloadCount +
                ", deleted=" + deleted +
                ", createTime=" + getCreateTime() +
                ", createBy='" + getCreateBy() + '\'' +
                '}';
    }
}

File: rookie-ui/src/types/api/knowhub/resource.ts
/**
 * 文件作用：
 * 定义资源管理模块对接后端所需的接口类型，
 * 与后端 ResourceVo / ResourceQuarry / ResourceReviewVo / ResourceReviewLogVo 对齐，
 * 供 api 层和页面层统一消费。
 * 关键约定：
 * - 时间字段沿用后端 ResourceVo 的 String 形式（BeanUtil 复制 Date→String）。
 * - resourceType 取值见字典 resource_type（FILE/LINK）；status 见字典 resource_status；
 *   reviewStatus 见字典 review_status（复用，NONE/PENDING/APPROVED/REJECTED）；
 *   action 见字典 review_action（SUBMIT/APPROVE/REJECT/REVOKE/PUBLISH，博客+资源共用）。
 * - 互动计数（likeCount/collectCount/ratingAvg/ratingCount）由后端聚合事实表回填，不冗余主表。
 * - hasLiked/hasCollected/myScore 为详情接口回填的当前用户态，管理台列表一般不消费。
 * - resourceCategoryId 为 -1 表示"其他"（前端硬编码约定，不查分类表）。
 */
import type { NormalizedPageResult, PageQueryParams } from '../system/common'

/**
 * 资源记录，与后端 ResourceVo 字段对齐。
 * FILE 类型：fileObjectId 关联 file_object，originalName/contentLength/contentType 由后端 join 带出。
 * LINK 类型：linkUrl 外部链接，linkIcon 图标（可空，前端可运行时拼 favicon 兜底）。
 */
export interface ResourceRecord {
  resourceId?: number
  authorId?: number
  resourceType: string
  resourceCategoryId?: number
  /** 分类名（后端 join 带出；-1=其他时为 null，前端硬编码展示"其他"） */
  categoryName?: string
  title: string
  summary?: string
  description?: string
  /** FILE 类型：关联 file_object.object_id */
  fileObjectId?: number
  /** FILE 类型：原始文件名（后端 join file_object 带出） */
  originalName?: string
  /** FILE 类型：文件大小字节（后端 join file_object 带出） */
  contentLength?: number
  /** FILE 类型：MIME 类型（后端 join file_object 带出） */
  contentType?: string
  /** LINK 类型：外部链接 URL */
  linkUrl?: string
  /** LINK 类型：图标 URL（可空） */
  linkIcon?: string
  status?: string
  reviewStatus?: string
  publishTime?: string
  /** 下载次数（仅 FILE 下载 +1，主表冗余） */
  downloadCount?: number
  /** 点赞数（事实表聚合回填） */
  likeCount?: number
  /** 收藏数（事实表聚合回填） */
  collectCount?: number
  /** 平均评分 0-5（事实表聚合回填） */
  ratingAvg?: number
  /** 评分数（事实表聚合回填） */
  ratingCount?: number
  hasLiked?: boolean
  hasCollected?: boolean
  /** 当前用户评分（详情接口回填，无则 0） */
  myScore?: number
  /** FILE 下载链接（详情接口按访问模式回填） */
  downloadUrl?: string
  createBy?: string
  /** 作者昵称（后端 join sys_user on author_id 带出） */
  authorNickname?: string
  createTime?: string
  updateBy?: string
  updateTime?: string
}

/**
 * 资源列表查询参数，与后端 ResourceQuarry + 分页参数对齐。
 * beginTime / endTime 由页面把日期范围控件拆成两个字段回传后端。
 */
export interface ResourceListQuery extends Partial<PageQueryParams> {
  title?: string
  resourceType?: string
  resourceCategoryId?: number
  status?: string
  reviewStatus?: string
  createBy?: string
  beginTime?: string
  endTime?: string
}

export type ResourcePageResult = NormalizedPageResult<ResourceRecord>

/**
 * 资源审核入参，与后端 ResourceReviewVo 对齐。
 * pass=true 通过 → PUBLISHED；pass=false 驳回 → REJECTED（advice 必填）。
 */
export interface ResourceReviewPayload {
  resourceId: number
  pass: boolean
  advice?: string
}

/**
 * 资源审核流水记录，与后端 ResourceReviewLogVo 字段对齐。
 * action 取值见字典 review_action（SUBMIT/APPROVE/REJECT/REVOKE/PUBLISH，博客+资源共用）。
 * operator 为操作人用户名快照，operatorNickname 由后端 join sys_user 带出便于直接展示。
 * role 为审核业务身份（AUTHOR作者/REVIEWER审核员/SYSTEM系统直通），按动作类型定非系统角色。
 */
export interface ResourceReviewLogRecord {
  reviewLogId?: number
  resourceId?: number
  action: string
  operatorId?: number
  operator: string
  operatorNickname?: string
  role: string
  advice?: string
  createTime?: string
}

/**
 * 资源分类扁平记录，与后端 ResourceCategoryVo 字段对齐。
 * parent_id=0 表示顶级；status:0禁1启。
 */
export interface ResourceCategoryRecord {
  categoryId?: number
  parentId?: number
  categoryName: string
  sort?: number
  status?: number
  createBy?: string
  createTime?: string
  updateBy?: string
  updateTime?: string
}

/**
 * 资源分类树形记录，与后端 ResourceCategoryTreeVo 对齐，供前端 el-tree 渲染。
 * 前端约定 -1=其他（不在分类树内，由前端硬编码加一个"其他"虚拟节点）。
 */
export interface ResourceCategoryTreeNode {
  categoryId?: number
  parentId?: number
  categoryName: string
  sort?: number
  status?: number
  children?: ResourceCategoryTreeNode[]
}
